### DISCUSSION TITLE
Orchard core, open Id, keycloack - how to

### DISCUSSION THREAD

**Discussion by vitalybrandes** [OrchardCMS/OrchardCore#7714] [Q&A]
Can I configure orchard core openid module to work with keycloack?
I am with orchard for last 6-7 month, but to.be honest, open I'd, still kind of the black box for me.
I would.liem to use keycloack to.manage all my identity and auth,
But how???
Also, doni need to create copy of that users in orchard? 🤔

**Comment by Piedone** [comment]
There are several video and written tutorials on configuring the Open ID module on the web, where exactly are you stuck?

**Reply by vitalybrandes** [reply]
@Piedone 
I do know how  integrate open id module in to Orchard core, but, i thinking about integration of external (Keycloack) as identity  and auth provider for all company app's.
I am not sure if i need to create users localy  (if i am using access by role) or open id map it for existing session?
Also, if i do need, can i "copy" password? do i need to create password?
How i create local user that mapped to external?

thank you 👍🏻

**Reply by MichaelPetrinolis** [reply]
You can follow the https://docs.orchardcore.net/en/dev/docs/guides/azuread-integration/ tutorial, and in place of Azure use keycloak. Just configure the OpenId client to connect to keycloak instead of AzureAD provider. The users are created locally and mapped to orchard roles based on the roles they have in your IdP. They don't get a password, so they cannot login locally. 
Then after you have tested your map script and everything works as expected (check that you have mapped the admin role), you can enable the 'Use external provider for login' and 'Disable local password login'. This way OrchardCore will challenge keycloak every time that you need to access admin and authenticate you.

**Reply by vitalybrandes** [reply]
@MichaelPetrinolis 
Thank you,
i assume that i can use both, external and internal usernames? right?

Please confirm that i understand,
once user mapped and login, orchard will create same local user with same ORCHARD CORE roles?
Any permission based view in my app will be based on ORCHARD CORE roles?
Last one, as you are creator , what name should i use instead AzureAD provider? OpenID?

**Reply by MichaelPetrinolis** [reply]
You should define the Orchard Core roles and permissions, based on your app needs. Then when a user is authenticated from your IdP will be created locally if does not exist based on the email claim received from the IdP. Then the user's roles are updated based on the claims. This logic is defined in the script or in a Custom Module. The Application Permissions are defined in the Orchard Core roles.

**Reply by vitalybrandes** [reply]
Thanks, 
ill try.

**Reply by vitalybrandes** [reply]
@MichaelPetrinolis  what name should i use instead AzureAD provider? OpenID?

**Reply by MichaelPetrinolis** [reply]
You should use the OpenIdConnect Client feature and configure it for use with your IdP.

**Reply by vitalybrandes** [reply]
Sorry for my misunderstanding but, what i need to user here instead azure ?

![image](https://user-images.githubusercontent.com/61165619/100726884-31fee500-33ce-11eb-80d0-2f4485957bfe.png)

**Reply by MichaelPetrinolis** [reply]
you can use this statement in script to figure it out 
```javascript
log("Warning", "Provider is {loginProvider}", context.loginProvider);
```
